Ejercicio 1 - Escribe una función que reciba una cadena de texto como parámetro y devuelva un diccionario con las frecuencias de cada letra en la cadena. Los espacios no deben ser considerados.

In [ ]:
def obtain_letters_frequency(string):
    """Función que recibe un string y devuelve un diccionario con la frecuencia de cada caracter (omitiendo los espacios)

    Args:
        string (str): Cualquier cadena de texto

    Returns:
        frequencies (dict): En cada par, la clave es el caracter y el valor es la frecuencia del mismo dentro del string
    """
    frequencies = {}
    for i in range(len(string)):
        frequencies_keys = frequencies.keys()
        if string[i] == " ":
            continue        
        if string[i] in frequencies_keys:
            frequencies[string[i]] += 1
        else:
            frequencies[string[i]] = 1
    return frequencies

print(obtain_letters_frequency("Hola, me llamo Paco"))

{'H': 1, 'o': 3, 'l': 3, 'a': 3, ',': 1, 'm': 2, 'e': 1, 'P': 1, 'c': 1}


1. He declarado un diccionario vacío.
2. Posteriormente, he creado un bucle que recorre el string caracter por caracter (usando la función range):
    - En cada iteración, se genera la variable de claves del diccionario, lo que permite comprobar a continuación si ese caracter ya está incluido en el diccionario (en cuyo caso se suma 1 a su frecuencia) o si no lo está (se crea con el valor 1). 
    - Además, como primera condición, he establecido que si el caracter es un espacio, directamente se pasa a la siguiente iteración mediante la palabra clave continue, ya que los espacios no nos interesan en este caso. 
    - Por otro lado, en la segunda condición he utilizado la palabra clave "in", que funciona como el método .includes de JavaScript, es decir, analiza si en el objeto está contenido un elemento. Esto no sé si salía en el temario, pero he tenido que buscarlo.
3. Al final de todo, la función devuelve el diccionario creado.

In [ ]:
def obtain_letters_frequency2(string):
    """Función que recibe un string y devuelve un diccionario con la frecuencia de cada caracter (omitiendo los espacios)

    Args:
        string (str): Cualquier cadena de texto

    Returns:
        frequencies_ordered (dict): En cada par, la clave es el caracter y el valor es la frecuencia del mismo dentro del string. Además, este diccionario está ordenado por orden alfabético en sus claves.
    """
    frequencies = {}
    string_en_minusculas = string.lower()
    frequencies_keys = frequencies.keys()
    for char in string_en_minusculas:
        if char < "a" or char > "z":
            continue        
        if char in frequencies_keys:
            frequencies[char] += 1
        else:
            frequencies[char] = 1
    frequencies_ordered = dict(sorted(frequencies.items()))
    return frequencies_ordered

print(obtain_letters_frequency2("Hola, me llamo Paco"))

{'a': 3, 'c': 1, 'e': 1, 'h': 1, 'l': 3, 'm': 2, 'o': 3, 'p': 1}


Cambios hechos:

1. Aplicar el bucle directamente sobre el string (en un principio no sabía que se podía recorrer strings con la estructura de bucles de Python, pensaba que sólo funcionaba con estructuras de datos como listas, tuplas, diccionarios y sets)
2. Crear la lista de claves del diccionario fuera del bucle hace que no se tenga que crear en cada iteración y funcione de la misma manera igualmente
3. Transformar el string en minúsculas para que no influya la mayúscula o la minúscula a la hora de contar la frecuencia de una letra
4. Atacar a cualquier caracter que no sea una letra (ya que en el enunciado se habla específicamente de letras). Esto lo hago haciendo que pase a la siguiente iteración sin hacer nada en caso de que el caracter sea "menor" que a y "mayor" que z (respecto a los rangos de ASCII)
5. Ordenar por orden alfabético las claves del diccionario de frecuencias, ordenando mediante el método sorted los pares de ítems del mismo, lo cual genera una lista de tuplas en la que cada tupla es un par clave-valor, y transformando esa lista en un diccionario posteriormente.


Ejercicio 2 - Dada una lista de números, obtén una nueva lista con el doble de cada valor. Usa la función map().

In [7]:
numbers_list = [2, 5, 6, 9, 10, 15, 24, 28, 35, 39]

double_numbers = list(map(lambda x: x*2, numbers_list))

print(double_numbers)

[4, 10, 12, 18, 20, 30, 48, 56, 70, 78]


1. He creado la función map utilizando una función lambda (las acabo de trabajar en la teoría) e incluyendo la lista de números
2. He transformado el retorno en una lista

In [8]:
double_numbers2 = [elemento * 2 for elemento in numbers_list]

print(double_numbers2)

[4, 10, 12, 18, 20, 30, 48, 56, 70, 78]


Alternativa con la list comprehension en lugar de la función map

Ejercicio 3 - Escribe una función que tome una lista de palabras y una palabra objetivo como parámetros. La función debe devolver una lista con todas las palabras de la lista original que contengan la palabra objetivo.

In [4]:
words = ["camino", "caminante", "camiseta", "cámara", "bicicleta", "automóvil", "autopista", "autonomía", "computadora", "composición", "avión", "avioneta", "aviso", "televisión", "telepatía", "lluvia", "llavero", "llanura", "playa", "plátano"]

words_with_auto = list(filter(lambda x: "auto" in x, words))
words_with_tele = list(filter(lambda x: "tele" in x, words))
words_with_vi = list(filter(lambda x: "vi" in x, words))

print(words_with_auto)
print(words_with_tele)
print(words_with_vi)

['automóvil', 'autopista', 'autonomía']
['televisión', 'telepatía']
['automóvil', 'avión', 'avioneta', 'aviso', 'televisión', 'lluvia']


1. He usado el método filter, incluyendo una función lambda en la que aplico la palabra clave "in" para analizar si el fragmento elegido está incluido en cada palabra de la lista.
2. Después, he transformado el objeto devuelto en una lista.

Lo he hecho con varios fragmentos

In [5]:
words_with_auto2 = [elemento for elemento in words if "auto" in elemento]

print(words_with_auto2)

['automóvil', 'autopista', 'autonomía']


Alternativa con list comprehension en lugar de la función filter

In [8]:
def filter_words(word, words_list):
    """Función que recibe una palabra o fragmento y una lista y filtra la lista, manteniendo sólo las palabra que incluyen el fragmento recibido como argumento

    Args:
        word (string): Fragmento que servirá como filtro
        words_list (list): Lista de palabras que se filtrará

    Returns:
        list: Lista filtrada
    """
    words_list_filtered = list(filter(lambda x: word in x, words_list))
    return words_list_filtered

print(filter_words("auto", words))
print(filter_words("tele", words))
print(filter_words("vi", words))

['automóvil', 'autopista', 'autonomía']
['televisión', 'telepatía']
['automóvil', 'avión', 'avioneta', 'aviso', 'televisión', 'lluvia']


Por último, he generado la función para aplicarla siempre que se quiera, la cual requiere como parámetros la palabra o fragmento y la lista objetivo

Ejercicio 4 - Genera una función que calcule la diferencia entre los valores de dos listas. Usa la función map().

In [15]:
def calculate_subtraction(first_list, second_list):
    """Función que recibe dos listas y resta los elementos de la segunda a los de la primera, emparejados por su índice
    Args:
        first_list (list): Lista de números enteros o decimales
        second_list (list): Lista de números enteros o decimales

    Returns:
        list: Lista de diferencias
    """
    subtraction = list(map(lambda pair: pair[0]-pair[1], zip(first_list, second_list)))
    return subtraction

print(calculate_subtraction([4, 6, 12], [2, 3, 4]))
print(calculate_subtraction([5, 3, 2], [5, 4, -5]))

[2, 3, 8]
[0, -1, 7]


En esta función hay tres niveles. Los analizo desde dentro hasta fuera:

1. He usado la función zip() para generar una lista de tuplas que emparejase a los elementos de la primera y la segunda lista según su índice.
2. He aplicado la función map() con una función lambda que devuelve el resultado de restar el elemento de la primera lista menos el elemento de la segunda (que son el primer y el segundo elemento de la tupla respectivamente), esto a lo largo de todos los índices de ambas listas (siendo cada índice una tupla).
3. He transformado el objeto resultante de la función map() en una lista a través de la palabra clave list() y lo he guardado todo en la variable substraction, que es la que se devuelve al final de la función

In [22]:
def calculate_subtraction2(first_list, second_list):
    """Función que recibe dos listas y resta los elementos de la segunda a los de la primera, emparejados por su índice
    Args:
        first_list (list): Lista de números enteros o decimales
        second_list (list): Lista de números enteros o decimales

    Returns:
        list: Lista de diferencias
    """
    subtraction = list(map(lambda x, y: x - y, first_list, second_list))
    return subtraction

print(calculate_subtraction2([4, 6, 12], [2, 3, 4]))
print(calculate_subtraction2([5, 3, 2], [5, 4, -5]))

[2, 3, 8]
[0, -1, 7]


Alternativa sin usar la función zip(), incluyendo las dos listas como argumentos directamente en la función map() (no sabía que esto se podía hacer), y creando una variable para cada lista en la función lambda()

In [16]:
def calculate_subtraction3(first_list, second_list):
    """Función que recibe dos listas y resta los elementos de la segunda a los de la primera, emparejados por su índice
    Args:
        first_list (list): Lista de números enteros o decimales
        second_list (list): Lista de números enteros o decimales

    Returns:
        list: Lista de diferencias
    """
    subtraction = [x - y for x, y in zip(first_list, second_list)]
    return subtraction

print(calculate_subtraction3([4, 6, 12], [2, 3, 4]))
print(calculate_subtraction3([5, 3, 2], [5, 4, -5]))

[2, 3, 8]
[0, -1, 7]


Alternativa sin usar la función map() (no es lo que pide el enunciado pero por rizar el rizo algo más), generando una list comprehension y utilizando en ella la función zip como en la primera resolución, para crear una lista de tuplas que incluyen las parejas de elementos según su índice.

Ejercicio 5 - Escribe una función que tome una lista de números como parámetro y un valor opcional nota_aprobado (por defecto 5). La función debe calcular la media de los números en la lista y determinar si la media es mayor o igual que nota_aprobado. Si es así, el estado será "aprobado"; de lo contrario, "suspenso". La función debe devolver una tupla que contenga la media y el estado.

In [15]:
def determine_passed(exams, nota_aprobado = 5):
    """Función que recibe una lista de números, calcula su media y comprueba si está por encima del valor por defecto nota_aprobado. Después, devuelve una tupla con la media y el estado (aprobado si es por encima o igual que 5 o suspenso si está por debajo)

    Args:
        exams (list): lista de notas a las que se les aplicará la media
        nota_aprobado (int): Nota límite para determinar el aprobado o el suspenso. Por defecto es 5.

    Returns:
        tuple: Tupla que agrupa la media y el estado (aprobado o suspenso)
    """
    total = 0
    for exam in exams:
        total += exam
    avg = total / len(exams)
    if avg >= nota_aprobado:
        state = "aprobado"
    else:
        state = "suspenso"
    result = (avg, state)
    return result
    

print(determine_passed([2, 7, 8, 6, 4, 9, 5, 10]))
        

(6.375, 'aprobado')


Para resolver el ejercicio, he seguido los siguientes pasos:

1. Para empezar, he calculado la suma total de todas las notas que incluya la lista. Para ello, he hecho un bucle for y he utilizado el operador de asignación combinada de la suma.
2. Después, he calculado la media de las notas, aprovechando el total anterior y usando el método len() para saber la cantidad de elementos
3. He utilizado el condicional if else para reflejar el estado de la media (aprobada o suspensa), en relación a la variable por defecto nota_aprobado (definida al principio de la función gracias al igual)
4. Por último, he creado la tupla result que incluía las variables avg y state, definidas anteriormente, y la he asignado como retorno de la función

In [18]:
def determine_passed2(exams, nota_aprobado = 5):
    """Función que recibe una lista de números, calcula su media y comprueba si está por encima del valor por defecto nota_aprobado. Después, devuelve una tupla con la media y el estado (aprobado si es por encima o igual que 5 o suspenso si está por debajo)

    Args:
        exams (list): lista de notas a las que se les aplicará la media
        nota_aprobado (int): Nota límite para determinar el aprobado o el suspenso. Por defecto es 5.

    Returns:
        tuple: Tupla que agrupa la media y el estado (aprobado o suspenso)
    """
    total = sum(exams)
    avg = total / len(exams)
    state = "aprobado" if avg >= nota_aprobado else "suspenso"
    result = (round(avg, 2), state)
    return result
    

print(determine_passed2([2, 7, 8, 6, 4, 9, 5, 10]))

(6.38, 'aprobado')


Alternativa aplicando varias estrategias para simplificar el código:
1. La función sum() para sumar directamente los elementos de una lista en lugar de pasar por el bucle for
2. Un ternario en el estado para reducir el condicional if else que determina si la media está aprobada o suspensa.

Además, he redondeado la media a dos decimales (mediante el método round()) después de compararla, para que no influya en el estado pero sí se muestren pocos decimales en la tupla final.

In [ ]:
def determine_passed3(exams, nota_aprobado = 5):
    """Función que recibe una lista de números, calcula su media y comprueba si está por encima del valor por defecto nota_aprobado. Después, devuelve una tupla con la media y el estado (aprobado si es por encima o igual que 5 o suspenso si está por debajo)

    Args:
        exams (list): lista de notas a las que se les aplicará la media
        nota_aprobado (int): Nota límite para determinar el aprobado o el suspenso. Por defecto es 5.

    Returns:
        tuple: Tupla que agrupa la media y el estado (aprobado o suspenso)
    """
    try:
        total = sum(exams)
        avg = total / len(exams)
    except ZeroDivisionError:
        print(f"La lista de notas está vacía")
    else:
        state = "aprobado" if avg >= nota_aprobado else "suspenso"
        result = (round(avg, 2), state)
        return result

print(determine_passed3([2, 7, 8, 6, 4, 9, 5, 10]))
print(determine_passed3([]))

(6.38, 'aprobado')
La lista de notas está vacía
None


Para rizar un poco más el rizo, he generado una estructura de manejo de errores con try, except y else para prevenir aquellos casos en los que la lista está vacía, lo cual quedaría plasmado en un ZeroDivisionError

Ejercicio 6 - Escribe una función que calcule el factorial de un número de manera recursiva.

In [15]:
def calculate_factorial(number):
    """Función que recibe un número entero y devuelve el factorial de dicho número

    Args:
        number (int): número entero al que se le calculará el factorial

    Returns:
        int: factorial del número recibido
    """
    anterior = number - 1
    if number > 0:
        return number * calculate_factorial(anterior)
    else:
        return 1
    
print(calculate_factorial(5))

120


Este ejercicio me ha costado de manera recursiva, he tenido que pedirle una pista al ChatGPT. Aun así, en la última resolución lo dejo como se me ocurrió nada más leerlo (sin función recursiva en ese caso). Aquí la explicación:
1. En primer lugar, he generado el número anterior (que nos servirá para provocar el bucle)
2. Después, he añadido un condicional que multiplica el número por esta misma función con el número anterior (recursividad)
3. Además, en el else (cuando se llega a 0), se devuelve un 1 para no generar un error en la multiplicación (aquí es donde estaba atrancado porque estaba recibiendo un error constantemente y no sabía de dónde salía)

En definitiva el proceso que sigue el programa es el siguiente:
Vuelta 1: 5 x (4!)
Vuelta 2: 5 x (4 x (3!))
Vuelta 3: 5 x (4 x (3 x (2!)))
Vuelta 4: 5 x (4 x (3 x (2 x (1!))))
Vuelta 5: 5 x (4 x (3 x (2 x (1)))) que, en definitiva, es 5 x 4 x 3 x 2 x 1 = 120


In [18]:
def calculate_factorial_one_line(number):
    """Función que recibe un número entero y devuelve el factorial de dicho número

    Args:
        number (int): número entero al que se le calculará el factorial

    Returns:
        int: factorial del número recibido
    """
    return number * calculate_factorial(number - 1) if number > 0 else 1

print(calculate_factorial_one_line(5))


120


Alternativa recursiva en una sola línea con un ternario

In [19]:
def calculate_factorial2(number):
    """Función que recibe un número entero y devuelve el factorial de dicho número

    Args:
        number (int): número entero al que se le calculará el factorial

    Returns:
        int: factorial del número recibido
    """
    factorial = 1
    for i in range(number):
        factorial *= i + 1
    return factorial

print(calculate_factorial2(5))
print(calculate_factorial2(12))

120
479001600


Esta no es una función recursiva pero fue la primera que se me ocurrió al momento de leer el enunciado y quería incluirla. Lo he conseguido mediante la función range() y el operador de asignación combinado de la multiplicación.

Ejercicio 7 - Genera una función que convierta una lista de tuplas a una lista de strings. Usa la función map().

In [28]:
def transform_tuples_into_strings(tuples_list):
    """Función que recibe una lista de tuplas y devuelve una lista de strings (uniendo dichas tuplas)

    Args:
        tuples_list (list): Lista cuyos elementos son tuplas

    Returns:
        list: Lista con las tuplas transformadas en strings
    """
    string_list = list(map(lambda x: f"{x[0]} {x[1]}", tuples_list))
    return string_list

pairs = [
    ("Coche", "rojo"),
    ("Gato", "negro"),
    ("Córdoba,", "España"),
    ("Ordenador", "portátil"),
    ("Fernando", "Alonso")
]

print(transform_tuples_into_strings(pairs))

['Coche rojo', 'Gato negro', 'Córdoba, España', 'Ordenador portátil', 'Fernando Alonso']


Esta es la primera solución que se me ocurre, aunque sólo sería viable si todas las tuplas tienen exactamente dos elementos.
1. He usado la función map() con una función lambda en su interior que unía el primer elemento de la tupla con el segundo en un string.
2. He transformado el objeto resultante en una lista y la he retornado

In [34]:
def transform_tuples_into_strings2(tuples_list):
    """Función que recibe una lista de tuplas y devuelve una lista de strings (uniendo dichas tuplas)

    Args:
        tuples_list (list): Lista cuyos elementos son tuplas

    Returns:
        list: Lista con las tuplas transformadas en strings
    """
    string_list = list(map(lambda x: " ".join(x), tuples_list))
    return string_list

mixed_tuples = [
    ("Paco", "Canovaca", "Cárdenas"),
    ("Coche", "Rojo"),
    ("Python", "es", "mi", "lenguaje", "favorito"),
    ("Fernando", "Alonso", "es", "piloto"),
    ("Córdoba,", "Andalucía,", "España")
]

print(transform_tuples_into_strings2(mixed_tuples))

['Paco Canovaca Cárdenas', 'Coche Rojo', 'Python es mi lenguaje favorito', 'Fernando Alonso es piloto', 'Córdoba, Andalucía, España']


Revisando los apuntes para encontrar una estrategia aplicable a cualquier número de elementos por tupla he llegado al método .join(), que había olvidado completamente. Así es aún más sencilla mi función y además puedo usarla sin importar cuántos elementos haya en cada tupla de mi lista.

El proceso es igual, sólo cambia en la función lambda el método join en lugar de unirlo yo manualmente.

Ejercicio 8 - Escribe un programa que pida al usuario dos números e intente dividirlos. Si el usuario ingresa un valor no numérico o intenta dividir por cero, maneja esas excepciones de manera adecuada y muestra un mensaje indicando si la división fue exitosa o no.

In [ ]:
def divide_two_numbers():
    try:
        num1 = float(input("Inserte el dividendo: "))
        num2 = float(input("Inserte el divisor: "))
        division = num1 / num2
        division_rounded = round(division, 2)
    except ValueError:
        print("Ambos valores deben ser números enteros o decimales")
    except ZeroDivisionError:
        print("El divisor debe ser distinto de 0")
    else:
        print(f"¡División exitosa! El resultado es {division_rounded:g}")

divide_two_numbers()

¡División exitosa! El resultado es 5


En este ejercicio, he utilizado la estructura de manejo de errores Try, Except (con dos tipos de error diferente) y Else:
1. En primer lugar, en el try he incluido los inputs para el dividendo y el divisor, la división de los dos números y el pequeño detalle de redondear los decimales a 2 (usado en un ejercicio anterior)
2. Después, he añadido la excepción de ValueError para los casos en que los valores añadidos en los inputs no sean números, lo cual no podría ser transformado a float.
3. Luego, he incluido otra excepción de ZeroDivisionError para cuando el divisor sea un 0.
4. Por último, si no se produce ningún error, el else incluye la impresión por consola del mensaje con el resultado de la división. Aquí he añadido otro pequeño detalle, esta vez nuevo y recién descubierto buscando en internet, que es el método :g, el cual hace que se eliminen los 0 innecesarios en los decimales. Es decir, si el resultado no tuviese ningún decimal o sólo uno, se mostraría únicamente ese decimal. Por ejemplo, en un resultado 5, se muestra así tal cual en lugar de 5.00

Ejercicio 9 - Escribe una función que tome una lista de nombres de mascotas como parámetro y devuelva una nueva lista excluyendo ciertas mascotas prohibidas en España. La lista de mascotas a excluir es ["Mapache", "Tigre", "Serpiente Pitón", "Cocodrilo", "Oso"]. Usa la función filter().

In [8]:
def filter_pets(*pets):
    forbidden_pets = ["Mapache", "Tigre", "Serpiente Pitón", "Cocodrilo", "Oso"]
    filtered_pets = list(filter(lambda x: x not in forbidden_pets, pets))
    return filtered_pets

pets_list = ["Perro", "Gato", "Mapache", "Cocodrilo", "Hámster", "Oso"]

print(f"Las mascotas permitidas son: {filter_pets(pets_list)}")

Las mascotas permitidas son: [['Perro', 'Gato', 'Mapache', 'Cocodrilo', 'Hámster', 'Oso']]
